File: AngularJS/src/org/angular2/entities/Angular2DirectiveSelectorSymbol.java
// Copyright 2000-2021 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.angular2.entities;

import com.intellij.find.usages.api.SearchTarget;
import com.intellij.find.usages.api.UsageHandler;
import com.intellij.javascript.web.codeInsight.html.WebSymbolsHtmlAdditionalContextProvider;
import com.intellij.javascript.web.symbols.WebSymbol;
import com.intellij.lang.documentation.DocumentationTarget;
import com.intellij.lang.javascript.psi.ecma6.TypeScriptClass;
import com.intellij.model.Pointer;
import com.intellij.navigation.NavigationRequest;
import com.intellij.navigation.NavigationTarget;
import com.intellij.navigation.TargetPresentation;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.TextRange;
import com.intellij.pom.Navigatable;
import com.intellij.psi.PsiElement;
import com.intellij.psi.search.SearchScope;
import com.intellij.psi.util.PsiTreeUtil;
import com.intellij.refactoring.rename.api.RenameTarget;
import com.intellij.util.containers.ContainerUtil;
import com.intellij.xml.XmlElementDescriptor;
import org.angular2.entities.impl.TypeScriptElementDocumentationTarget;
import org.angular2.web.Angular2Symbol;
import org.angular2.web.Angular2SymbolOrigin;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.regex.Pattern;

import static com.intellij.javascript.web.codeInsight.html.WebSymbolsHtmlAdditionalContextProvider.getHtmlNSDescriptor;
import static com.intellij.javascript.web.symbols.WebSymbolsUtils.createPsiRangeNavigationItem;
import static org.angular2.Angular2DecoratorUtil.getClassForDecoratorElement;
import static org.angular2.web.Angular2WebSymbolsAdditionalContextProvider.KIND_NG_DIRECTIVE_ATTRIBUTE_SELECTORS;
import static org.angular2.web.Angular2WebSymbolsAdditionalContextProvider.KIND_NG_DIRECTIVE_ELEMENT_SELECTORS;

public class Angular2DirectiveSelectorSymbol implements Angular2Symbol, SearchTarget, RenameTarget {

  private static final Pattern TAG_NAME_PATTERN = Pattern.compile("[a-zA-Z][a-zA-Z0-9-]*");
  private static final Pattern ATTRIBUTE_NAME_PATTERN = Pattern.compile("[^\\p{Space}\"'>/=\\p{Cntrl}]+");

  private final Angular2DirectiveSelectorImpl myParent;
  private final TextRange myRange;
  private final String myName;
  private final String myElementSelector;
  private final boolean myIsElement;

  public Angular2DirectiveSelectorSymbol(@NotNull Angular2DirectiveSelectorImpl parent,
                                         @NotNull TextRange range,
                                         @NotNull String name,
                                         @Nullable String elementSelector,
                                         boolean isElement) {
    myParent = parent;
    myRange = range;
    myName = name;
    myElementSelector = elementSelector;
    myIsElement = isElement;
  }

  @Override
  public @NotNull String getName() {
    return myName;
  }

  @Nullable
  @Override
  public Priority getPriority() {
    return Priority.LOWEST;
  }

  @NotNull
  @Override
  public DocumentationTarget getDocumentationTarget() {
    var clazz = PsiTreeUtil.getContextOfType(getSource(), TypeScriptClass.class);
    if (clazz == null) {
      return Angular2Symbol.super.getDocumentationTarget();
    }
    return new TypeScriptElementDocumentationTarget(getName(), clazz);
  }

  @Nullable
  @Override
  public PsiElement getPsiContext() {
    return myParent.getPsiParent();
  }

  public @NotNull PsiElement getSource() {
    return myParent.getPsiParent();
  }

  @Override
  public @NotNull Project getProject() {
    return getSource().getProject();
  }

  @NotNull
  @Override
  public Namespace getNamespace() {
    return Namespace.JS;
  }

  @NotNull
  @Override
  public String getKind() {
    return myIsElement ? KIND_NG_DIRECTIVE_ELEMENT_SELECTORS : KIND_NG_DIRECTIVE_ATTRIBUTE_SELECTORS;
  }

  @NotNull
  @Override
  public Pointer<Angular2DirectiveSelectorSymbol> createPointer() {
    var parent = myParent.createPointer();
    var range = myRange;
    var name = myName;
    var elementName = myElementSelector;
    var isElement = myIsElement;
    return () -> {
      var newParent = parent.dereference();
      return newParent != null ? new Angular2DirectiveSelectorSymbol(newParent, range, name, elementName, isElement) : null;
    };
  }

  @NotNull
  @Override
  public Origin getOrigin() {
    return new Angular2SymbolOrigin(this);
  }

  public boolean isDeclaration() {
    return !ContainerUtil.exists(getReferencedSymbols(), symbol -> !(symbol instanceof Angular2Symbol));
  }

  public List<WebSymbol> getReferencedSymbols() {
    var psiElement = getSource();
    var nsDescriptor = getHtmlNSDescriptor(psiElement.getProject());
    if (nsDescriptor != null) {
      if (isElementSelector()) {
        var elementDescriptor = nsDescriptor.getElementDescriptorByName(getName());
        if (elementDescriptor != null) {
          return Collections.singletonList(
            new WebSymbolsHtmlAdditionalContextProvider.HtmlElementDescriptorBasedSymbol(elementDescriptor, null));
        }
      }
      else {
        XmlElementDescriptor elementDescriptor = null;
        String tagName = myElementSelector;
        if (myElementSelector != null) {
          elementDescriptor = nsDescriptor.getElementDescriptorByName(myElementSelector);
        }
        if (elementDescriptor == null) {
          elementDescriptor = nsDescriptor.getElementDescriptorByName("div");
          tagName = "div";
        }
        if (elementDescriptor != null) {
          var attributeDescriptor = elementDescriptor.getAttributeDescriptor(getName(), null);
          if (attributeDescriptor != null) {
            return Collections.singletonList(
              new WebSymbolsHtmlAdditionalContextProvider.HtmlAttributeDescriptorBasedSymbol(attributeDescriptor, tagName));
          }
        }
      }
    }
    return Collections.singletonList(this);
  }

  public int getTextOffset() {
    return myParent.getPsiParent().getTextOffset() + myRange.getStartOffset();
  }

  public @NotNull TextRange getTextRangeInSource() {
    return myRange;
  }

  public boolean isElementSelector() {
    return myIsElement;
  }

  public boolean isAttributeSelector() {
    return !myIsElement;
  }

  @Override
  public String toString() {
    return (myIsElement ? "ElementDirectiveSelector" : "AttributeDirectiveSelector") + "<" + myName + ">";
  }

  @Override
  public @NotNull Collection<NavigationTarget> getNavigationTargets(@NotNull Project project) {
    return Collections.singletonList(new DirectiveSelectorSymbolNavigationTarget(this));
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    Angular2DirectiveSelectorSymbol symbol = (Angular2DirectiveSelectorSymbol)o;
    return myIsElement == symbol.myIsElement &&
           myParent.equals(symbol.myParent) &&
           myRange.equals(symbol.myRange) &&
           myName.equals(symbol.myName) &&
           Objects.equals(myElementSelector, symbol.myElementSelector);
  }

  @Override
  public int hashCode() {
    return Objects.hash(myParent, myRange, myName, myElementSelector, myIsElement);
  }

  @Override
  public @NotNull TargetPresentation getPresentation() {
    PsiElement parent = myParent.getPsiParent();
    TypeScriptClass clazz = getClassForDecoratorElement(parent);
    return TargetPresentation.builder(getName())
      .icon(getIcon())
      .locationText(parent.getContainingFile().getName())
      .containerText(clazz != null ? clazz.getName() : null)
      .presentation();
  }

  @NotNull
  @Override
  public UsageHandler<?> getUsageHandler() {
    return UsageHandler.createEmptyUsageHandler(getName());
  }

  @NotNull
  @Override
  public String getTargetName() {
    return getName();
  }

  @Nullable
  @Override
  public SearchScope getMaximalSearchScope() {
    return SearchTarget.super.getMaximalSearchScope();
  }

  @Nullable
  @Override
  public String validateName(@NotNull String name) {
    if (myIsElement) {
      return TAG_NAME_PATTERN.matcher(name).matches() ? null : name +" is not a valid HTML element name.";
    }
    return ATTRIBUTE_NAME_PATTERN.matcher(name).matches() ? null : name + " is not a valid HTML attribute name.";
  }

  private static class DirectiveSelectorSymbolNavigationTarget implements NavigationTarget {

    private final Angular2DirectiveSelectorSymbol mySymbol;

    private DirectiveSelectorSymbolNavigationTarget(Angular2DirectiveSelectorSymbol symbol) {
      mySymbol = symbol;
    }

    @Override
    public final @NotNull Pointer<? extends NavigationTarget> createPointer() {
      return Pointer.delegatingPointer(
        mySymbol.createPointer(),
        DirectiveSelectorSymbolNavigationTarget.class,
        DirectiveSelectorSymbolNavigationTarget::new
      );
    }

    @Override
    public @NotNull Navigatable getNavigatable() {
      return createPsiRangeNavigationItem(this, mySymbol.getSource(), mySymbol.myRange.getStartOffset());
    }

    @Override
    public @NotNull TargetPresentation getTargetPresentation() {
      return mySymbol.getPresentation();
    }

    @Override
    public @Nullable NavigationRequest navigationRequest() {
      return getNavigatable().navigationRequest();
    }
  }
}

File: AngularJS/src/org/angular2/entities/Angular2DirectiveSelectorSymbol.kt
// Copyright 2000-2023 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package org.angular2.entities

import com.intellij.find.usages.api.SearchTarget
import com.intellij.html.webSymbols.WebSymbolsHtmlQueryConfigurator
import com.intellij.html.webSymbols.WebSymbolsHtmlQueryConfigurator.Companion.getHtmlNSDescriptor
import com.intellij.lang.javascript.psi.ecma6.TypeScriptClass
import com.intellij.model.Pointer
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.NlsSafe
import com.intellij.openapi.util.TextRange
import com.intellij.platform.backend.documentation.DocumentationTarget
import com.intellij.platform.backend.presentation.TargetPresentation
import com.intellij.psi.PsiElement
import com.intellij.psi.util.contextOfType
import com.intellij.refactoring.rename.api.RenameTarget
import com.intellij.refactoring.rename.api.RenameValidationResult
import com.intellij.refactoring.rename.api.RenameValidator
import com.intellij.webSymbols.WebSymbol
import com.intellij.webSymbols.declarations.WebSymbolDeclaration
import com.intellij.webSymbols.utils.WebSymbolDeclaredInPsi
import com.intellij.xml.XmlElementDescriptor
import org.angular2.Angular2DecoratorUtil.getClassForDecoratorElement
import org.angular2.codeInsight.documentation.Angular2ElementDocumentationTarget
import org.angular2.web.Angular2Symbol
import org.angular2.web.Angular2WebSymbolsQueryConfigurator.Companion.KIND_NG_DIRECTIVE_ATTRIBUTE_SELECTORS
import org.angular2.web.Angular2WebSymbolsQueryConfigurator.Companion.KIND_NG_DIRECTIVE_ELEMENT_SELECTORS
import org.angularjs.AngularJSBundle
import java.util.*
import java.util.regex.Pattern

class Angular2DirectiveSelectorSymbol(private val myParent: Angular2DirectiveSelectorImpl,
                                      override val textRangeInSourceElement: TextRange,
                                      override val name: @NlsSafe String,
                                      private val myElementSelector: String?,
                                      val isElementSelector: Boolean) : Angular2Symbol, SearchTarget, RenameTarget, WebSymbolDeclaredInPsi {

  override val priority: WebSymbol.Priority
    get() = WebSymbol.Priority.LOWEST

  override val psiContext: PsiElement
    get() = myParent.psiParent

  override val sourceElement: PsiElement
    get() = myParent.psiParent

  override val project: Project
    get() = sourceElement.project

  override val namespace: String
    get() = WebSymbol.NAMESPACE_JS

  override val kind: String
    get() = if (isElementSelector) KIND_NG_DIRECTIVE_ELEMENT_SELECTORS else KIND_NG_DIRECTIVE_ATTRIBUTE_SELECTORS

  override val declaration: WebSymbolDeclaration?
    get() = super.declaration.takeIf { isDeclaration }

  val isDeclaration: Boolean
    get() = referencedSymbols.all { it is Angular2Symbol }

  val referencedSymbols: List<WebSymbol>
    get() {
      val psiElement = sourceElement
      val nsDescriptor = getHtmlNSDescriptor(psiElement.project)
      if (nsDescriptor != null) {
        if (isElementSelector) {
          val elementDescriptor = nsDescriptor.getElementDescriptorByName(name)
          if (elementDescriptor != null) {
            return listOf<WebSymbol>(WebSymbolsHtmlQueryConfigurator.HtmlElementDescriptorBasedSymbol(elementDescriptor, null))
          }
        }
        else {
          var elementDescriptor: XmlElementDescriptor? = null
          var tagName = myElementSelector
          if (myElementSelector != null) {
            elementDescriptor = nsDescriptor.getElementDescriptorByName(myElementSelector)
          }
          if (elementDescriptor == null) {
            elementDescriptor = nsDescriptor.getElementDescriptorByName("div")
            tagName = "div"
          }
          if (elementDescriptor != null) {
            val attributeDescriptor = elementDescriptor.getAttributeDescriptor(name, null)
            if (attributeDescriptor != null) {
              return listOf<WebSymbol>(WebSymbolsHtmlQueryConfigurator.HtmlAttributeDescriptorBasedSymbol(attributeDescriptor, tagName!!))
            }
          }
        }
      }
      return listOf<WebSymbol>(this)
    }

  val isAttributeSelector: Boolean
    get() = !isElementSelector

  override val presentation: TargetPresentation
    get() {
      val parent = myParent.psiParent
      val clazz = getClassForDecoratorElement(parent)
      return TargetPresentation.builder(name)
        .icon(icon)
        .locationText(parent.containingFile.name)
        .containerText(clazz?.name)
        .presentation()
    }

  override fun getDocumentationTarget(location: PsiElement?): DocumentationTarget =
    Angular2ElementDocumentationTarget.create(
      name, location,
      Angular2EntitiesProvider.getEntity(sourceElement.contextOfType<TypeScriptClass>(true)))
    ?: super<Angular2Symbol>.getDocumentationTarget(location)

  override fun createPointer(): Pointer<Angular2DirectiveSelectorSymbol> {
    val parent = myParent.createPointer()
    val range = textRangeInSourceElement
    val name = this.name
    val elementName = myElementSelector
    val isElement = isElementSelector
    return Pointer {
      val newParent = parent.dereference()
      if (newParent != null) Angular2DirectiveSelectorSymbol(newParent, range, name, elementName, isElement) else null
    }
  }

  override fun toString(): String {
    return (if (isElementSelector) "ElementDirectiveSelector" else "AttributeDirectiveSelector") + "<" + name + ">"
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other == null || javaClass != other.javaClass) return false
    val symbol = other as Angular2DirectiveSelectorSymbol?
    return isElementSelector == symbol!!.isElementSelector &&
           myParent == symbol.myParent &&
           textRangeInSourceElement == symbol.textRangeInSourceElement &&
           name == symbol.name &&
           myElementSelector == symbol.myElementSelector
  }

  override fun hashCode(): Int {
    return Objects.hash(myParent, textRangeInSourceElement, name, myElementSelector, isElementSelector)
  }

  override fun validator(): RenameValidator {
    return MyRenameValidator(isElementSelector)
  }

  private class MyRenameValidator(private val myIsElement: Boolean) : RenameValidator {

    override fun validate(newName: String): RenameValidationResult {
      if (myIsElement) {
        return if (TAG_NAME_PATTERN.matcher(newName).matches())
          RenameValidationResult.ok()
        else
          RenameValidationResult.invalid(
            AngularJSBundle.message("angularjs.refactoring.selector.invalid.html.element.name", newName))
      }
      return if (ATTRIBUTE_NAME_PATTERN.matcher(newName).matches())
        RenameValidationResult.ok()
      else
        RenameValidationResult.invalid(
          AngularJSBundle.message("angularjs.refactoring.selector.invalid.html.attribute.name", newName))
    }
  }

  companion object {

    private val TAG_NAME_PATTERN = Pattern.compile("[a-zA-Z][a-zA-Z0-9-]*")
    private val ATTRIBUTE_NAME_PATTERN = Pattern.compile("[^\\s\"'>/=\\p{Cntrl}]+")
  }
}
